### ISSUE TITLE
updating the binaries used in CI tests for the dependency `storm`

### ISSUE THREAD

**Issue by johnyf** [tulip-control/tulip-control#246]
On the branch `ci_update_download` I have updated the CI configuration and dependencies (this was motivated by `bintray.com` being discontinued, as explained in a section below). The only remaining error:

https://travis-ci.com/github/tulip-control/tulip-control/jobs/510638673#L15866

reads:

```
File "/home/travis/build/tulip-control/tulip-control/tests/stormpy_interface_test.py", line 25, in <module>
    from tulip.interfaces import stormpy as stormpy_int
  File "/home/travis/virtualenv/python3.9.1/lib/python3.9/site-packages/tulip/interfaces/stormpy.py", line 39, in <module>
    import stormpy
  File "/home/travis/build/tulip-control/tulip-control/stormpy-1.6.2/lib/stormpy/__init__.py", line 6, in <module>
    from . import core
ImportError: libboost_filesystem.so.1.65.1: cannot open shared object file: No such file or directory
```

This error appears to be from [`storm`](https://github.com/moves-rwth/storm), to which [`stormpy`](https://github.com/moves-rwth/stormpy) interfaces. `stormpy` is built on the CI run:

https://github.com/tulip-control/tulip-control/blob/ece84cdccb4deb4cb14b144df7e5b7f5697eb869/extern/get-stormpy.sh#L79

so I would expect that it links (if it links at all) to the version of `libboost_filesystem.so` that is present on `focal` (and thus not raise the above error. `stormpy` is built using [`pybind11`](https://github.com/pybind/pybind11), and the [`core`](https://github.com/moves-rwth/stormpy/tree/1.6.2/src/core) mentioned in the traceback above is a C++ extension that is [built with](https://github.com/moves-rwth/stormpy/blob/1.6.2/src/mod_core.cpp#L14) `pybind11`). That the error is from `storm` is confirmed by attempting to run

[`storm --version`](https://www.stormchecker.org/documentation/usage/running-storm.html#first-steps)

in the CI run, which [raises](https://travis-ci.com/github/tulip-control/tulip-control/jobs/510660367#L15483):

```
./storm/build/bin/storm: error while loading shared libraries: libboost_system.so.1.65.1: cannot open shared object file: No such file or directory
```

(`libboost_filesystem.so.1.65.1` was the dependence in the traceback from above, but that was from `stormpy`, which is linked to `storm` when built, so which libraries are loaded from `stormpy`, and in which order, can differ from which libraries are loaded, and in which order, when calling `storm --version`.)

Indeed, the dependence on `libboost_filesystem` from above is confirmed as follows after unpacking the archive downloaded from https://sourceforge.net/projects/cudd-mirror/files/storm.tar.xz/download and `cd` inside it:

```
> pwd
.../storm/build/lib
> strings * | ag libboost_filesystem
libboost_filesystem.so.1.65.1
> ag --search-binary libboost_filesystem
Binary file libstorm.so matches.
```

(Using [`string`](https://en.wikipedia.org/wiki/Strings_(Unix)) and [`ag`](https://github.com/ggreer/the_silver_searcher).)

(Boost is a [requirement of `storm`](https://www.stormchecker.org/documentation/obtain-storm/dependencies.html#boost).)

Also, the error does not seem to be from the dependency of `storm`, because [`ldd carl/libcarl.so.14.20` prints in the CI run](https://travis-ci.com/github/tulip-control/tulip-control/jobs/510660367#L2586-L2597):

```
linux-vdso.so.1 (0x00007ffe28fa5000)
libgmpxx.so.4 => /home/travis/build/tulip-control/tulip-control/carl/resources/lib/libgmpxx.so.4 (0x00007f6708d80000)
libgmp.so.10 => /home/travis/build/tulip-control/tulip-control/carl/resources/lib/libgmp.so.10 (0x00007f6708b0a000)
libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f6708af5000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f6708ad2000)
libcln.so.6 => /home/travis/build/tulip-control/tulip-control/carl/resources/lib/libcln.so.6 (0x00007f670875b000)
libginac.so.6 => /home/travis/build/tulip-control/tulip-control/carl/resources/lib/libginac.so.6 (0x00007f67082ba000)
libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f67080d9000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f6707f8a000)
libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f6707f6f000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f6707d7d000)
/lib64/ld-linux-x86-64.so.2 (0x00007f670924a000)
```

whereas [`ldd storm/build/lib/libstorm.so` prints in the CI run](https://travis-ci.com/github/tulip-control/tulip-control/jobs/510660367#L15461-L15482):

```
linux-vdso.so.1 (0x00007ffd11d93000)
libcarl.so.14.20 => /home/travis/build/tulip-control/tulip-control/carl/libcarl.so.14.20 (0x00007f5bb1a68000)
libginac.so.6 => /home/travis/build/tulip-control/tulip-control/carl/resources/lib/libginac.so.6 (0x00007f5bb15c9000)
libcln.so.6 => /home/travis/build/tulip-control/tulip-control/carl/resources/lib/libcln.so.6 (0x00007f5bb1252000)
libgmpxx.so.4 => /home/travis/build/tulip-control/tulip-control/carl/resources/lib/libgmpxx.so.4 (0x00007f5bb104b000)
libgmp.so.10 => /home/travis/build/tulip-control/tulip-control/carl/resources/lib/libgmp.so.10 (0x00007f5bb0dd3000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f5bb0da1000)
libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f5bb0bc0000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f5bb0a71000)
libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f5bb0a56000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f5bb0864000)
/lib64/ld-linux-x86-64.so.2 (0x00007f5bb3d42000)
libboost_filesystem.so.1.65.1 => not found
libboost_system.so.1.65.1 => not found
libz3.so.4 => /lib/x86_64-linux-gnu/libz3.so.4 (0x00007f5baf2c2000)
libglpk.so.40 => /lib/x86_64-linux-gnu/libglpk.so.40 (0x00007f5baefe3000)
libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f5baefdd000)
libcolamd.so.2 => /lib/x86_64-linux-gnu/libcolamd.so.2 (0x00007f5baefd2000)
libamd.so.2 => /lib/x86_64-linux-gnu/libamd.so.2 (0x00007f5baefc7000)
libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f5baefab000)
libltdl.so.7 => /lib/x86_64-linux-gnu/libltdl.so.7 (0x00007f5baefa0000)
libsuitesparseconfig.so.5 => /lib/x86_64-linux-gnu/libsuitesparseconfig.so.5 (0x00007f5baef9b000)
```

In particular, notice in the `ldd` output for `libstorm.so` the lines:

```
libboost_filesystem.so.1.65.1 => not found
libboost_system.so.1.65.1 => not found
```

So the above error is raised because the binary of `storm` has been linked to `libboost_filesystem.so.1.65.1`, which [is present on Ubuntu `bionic`](https://packages.ubuntu.com/bionic/libboost-filesystem-dev), but absent on Ubuntu `focal`. On Ubuntu `focal`, [`libboost-all-dev (1.71.0.0ubuntu2)` is available](https://packages.ubuntu.com/focal/libboost-filesystem-dev).

Currently, binaries for `storm` are downloaded from https://sourceforge.net/projects/cudd-mirror/files/storm.tar.xz/download (also for https://sourceforge.net/projects/cudd-mirror/files/carl.tar.xz/download, though I do not know whether that program will run on `focal`--from the first `ldd` output above it seems that it might).

Approaches to consider:
- updating the binaries at:  https://sourceforge.net/projects/cudd-mirror/files/
- uploading new binaries as "release" assets at:  https://github.com/tulip-control/data. The `data` repository content would be minimal, simply describing textually what version of the executables are built using which script from the `tulip-control` repository. This would avoid committing binaries in `git` (even though the purpose of the `data` repository could be regarded as a place for binaries, among other kinds of files).

It will be some time before I will have access to a machine where I can build newer versions of these binaries suitable for use on Travis CI (the CI builds will fail until the issue is addressed). In any case, the changes from branch `ci_update_download` could be merged now into the mainline branch of `tulip`. Then the remaining update for the CI tests to pass would be to change the `storm` binaries.

## About caching binaries of dependencies on Travis CI

Caching on Travis CI the results of building `stormpy`-related packages has been [considered before](https://github.com/tulip-control/tulip-control/pull/237#issuecomment-703171430). However, building these programs takes hours (https://github.com/tulip-control/tulip-control/pull/237#issuecomment-703168131, https://github.com/tulip-control/tulip-control/pull/237#issuecomment-720674642), so I do not know how feasible it is to build the binaries on Travis CI.

An issue with such builds on Travis CI would be [timeouts](https://docs.travis-ci.com/user/customizing-the-build/#build-timeouts) (10 minutes if no output is produced by a CI instance, and 50 minutes overall for each CI instance). It appears that [timeouts can be extended](https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received) with `travis_wait`. However, `travis_wait` changes the timeout related to producing output, not the overall timeout. I do not know whether there is any way to change the overall timeout.

In addition, Travis CI [caches expire after 45 days](https://docs.travis-ci.com/user/caching/#caches-expiration) (for `travis-ci.com`). So building caches of executables on Travis CI would not avoid periodic rebuilds.

## Changes to CI configuration on branch `ci_update_download`

In the CI environment where `tulip` is tested, `gr1c` was downloaded from Bintray:

https://github.com/tulip-control/tulip-control/blob/bb004422b575dccea8d19c33acfeb04b37c62a5a/.travis.yml#L73

[Bintray](https://bintray.com) [shut down on May 1st, 2021](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/). The CI setup script [raises](https://travis-ci.org/github/tulip-control/tulip-control/jobs/772690567#L2031) an error about the SHA-256 hash.

On the branch `ci_update_download` I changed `tulip`'s CI configuration file `.travis.yml` to download the file https://github.com/tulip-control/gr1c/releases/download/v0.13.0/gr1c-0.13.0-Linux_x86-64.tar.gz (and in doing so I also bumped the vesion of `gr1c` used in the tests from version [0.11.0](https://github.com/tulip-control/gr1c/releases/tag/v0.11.0) to version [0.13.0](https://github.com/tulip-control/gr1c/releases/tag/v0.13.0)). This change happened in `tulip` commit https://github.com/tulip-control/tulip-control/commit/f63d882b7c42da816d233d250c6453c67c644de7.

Using this change, the [CI raised an error](https://travis-ci.org/github/tulip-control/tulip-control/jobs/772877421#L2833) that seems to relate to the version of [GLIBC](https://en.wikipedia.org/wiki/GNU_C_Library):

> ValueError: invalid literal for int() with base 10: "gr1c: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by gr1c)"

The requirement of GLIBC 2.29 is confirmed using `readelf -s gr1c-0.13.0-Linux_x86-64/gr1c`, which starts with:

```
Symbol table '.dynsym' contains 63 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND
     1: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND free@GLIBC_2.2.5 (2)
     2: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND log2@GLIBC_2.29 (3)
```

(`greadelf` on macOS is [Readelf](https://en.wikipedia.org/wiki/Readelf), which can be installed on macOS by installing the [MacPorts](https://www.macports.org) package [`binutils`](https://ports.macports.org/port/binutils/summary), on Linux `ldd gr1c-0.13.0-Linux_x86-64/gr1c` is possible too).

So I changed the OS for the CI run to `focal` (Ubuntu 20.04), from `bionic` (Ubuntu 18.04). This [raised another error](https://travis-ci.org/github/tulip-control/tulip-control/jobs/772881862#L266):

> The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends $(travis_apt_get_options) install gfortran libatlas-base-dev liblapack-dev libgmp-dev libmpfr-dev graphviz libglpk-dev libboost-dev libboost-filesystem-dev libboost-program-options-dev libboost-regex-dev libboost-test-dev libeigen3-dev z3 libz3-dev python-z3 libhwloc-dev" failed and exited with 100 during .

This error is because the `apt` package [`python-z3`](https://packages.ubuntu.com/bionic/python-z3), which is available on `bionic`, is not available on `focal` (from the default package repository). Instead, the corresponding package on `focal` is [`python3-z3`](https://packages.ubuntu.com/focal/python3-z3). So I changed the CI configuration to install `python3-z3`.

(`tulip` now supports only Python 3, so installing `python3-z3` suffices. Even if Python 2 was still supported, and thus tested in CI, `python3-z3` is required only for testing the interface to `stormpy`. The package `stormpy` requires Python 3, so the tests for that interface were run only on Python 3. Thus, it would have sufficed to conditionally intall `python3-z3` only if the CI run was for Python 3. In any case, this conditional installation is not needed now.)

After these changes, and an update to fetch an archive of a more recent commit of `slugs` (which contains Python scripts updated to Python 3), the error from the `storm` binary remains, which is discussed above.

**Comment by slivingston** [comment]
@johnyf thanks for this detailed treatment. Regarding updating and uploading built files for `storm` and `carl`, another idea is to configure https://github.com/tulip-control/data for [Git LFS](https://git-lfs.github.com/), which causes a hash of specified files to be tracked in the Git repository while the file itself is uploaded to a Git LFS conforming server. GitHub provides one at no cost.

> It will be some time before I will have access to a machine where I can build newer versions of these binaries suitable for use on Travis CI

I can work on this, but likely not until later this month.

**Comment by johnyf** [comment]
Using Git LFS I think is a better approach, @slivingston thank you for suggesting it. Thank you for being available to work on updating the binaries.

**Comment by johnyf** [comment]
Addressed in commit b38b525aca9d918b4737c10a1e02625cba4a49fd, by uploading new binaries for `storm` and `carl` as release assets of a GitHub release associated to a tag of the repository https://github.com/tulip-control/data.

I noted in the README of the `data` repository that tags are by default considered temporary, and in the release page what that release is for, and that the release may be replaced in the future (https://github.com/tulip-control/data/releases/tag/stormpy_dependencies).

These changes are tested successfully using GitHub Actions in commit 7ececa4bcae51c85759e7678f706c160e0fc4fa5, and have been merged to the mainline branch. So now all tests pass.


## About LFS on GitHub

I think that Git LFS is a good solution for this issue. However, GitHub's LFS storage appears to have limitations that could limit CI testing, as described below. In summary:

- downloads of Git LFS files from GitHub are limited to 1 GB in total per month
  (for free accounts). Any Git LFS activity on forks also counts towards the
  root repository owner's quotas.

- deleting old files from GitHub's LFS storage requires deleting the entire
  repository (`https://github.com/tulip-control/data`), and recreating it.

In comparison, I am not aware of limits on file size of release assets, and as far as I know, deleting a release from a GitHub repository's releases suffices to delete the release assets associated with the deleted release.

(Sidenote: The closest thing in the `git` ecosystem appears to be `git annex dropunused --force`, using [`git-annex`](https://en.wikipedia.org/wiki/Git-annex).)


## GitHub LFS storage and bandwidth limitations

There are limitations on storage bandwidth of GitHub LFS per repository owner:
https://docs.github.com/en/github/managing-large-files/versioning-large-files/about-storage-and-bandwidth-usage
It seems to me that this limitation could imply that testing on GitHub Actions may be limited by bandwidth usage.


### GitHub LFS storage limitation

The `stormpy`-related binaries currently in the `data` repository release assets have total size 76 MB. The binaries that were previously downloaded from SourceForge and used in CI had total size 72 MB.

With a 1 GB limit on GitHub LFS storage, about 13 versions of these binaries can be stored. Although these files need to change only when either the CI environment changes (e.g., newer versions of `apt` packages), or the dependencies being built are updated, when newer versions are being created, a number of iterations may be needed until the CI tests pass. These iterations would probably exhaust the 1 GB space.


### GitHub LFS bandwidth limitation


#### GitHub LFS and `git` bandwidth

The bandwidth limitation could be restrictive on CI testing, because each push triggers a CI build, which will download the files (one option might be to cache the files on GitHub Actions, and these particular files appear to be the use case that matches the reason that caching exists on GitHub Actions, though cache is cleared every 45 days), will be counted towards the bandwidth limit:

> If you choose to include Git LFS objects in archives of your repository, every download of those archives will count towards bandwidth usage for your account. Each account receives 1 GB per month of bandwidth for free, and you can pay for additional usage.

https://docs.github.com/en/github/administering-a-repository/managing-repository-settings/managing-git-lfs-objects-in-archives-of-your-repository

There are months when more than 13 pushes occur in the repository `tulip-control`, so having such a limit would make the builds fail after the limit is reached:

> If you use more than 1 GB of bandwidth per month without purchasing a data pack, Git LFS support is disabled on your account until the next month.

https://docs.github.com/en/github/managing-large-files/versioning-large-files/about-storage-and-bandwidth-usage#bandwidth-quota


#### GitHub LFS and bandwidth of repository downloads

There is an option in GitHub repository settings (in this case for the settings of the repository `data`) to include Git LFS files in repository archives (ZIP files and tarballs) that GitHub makes available for download:
https://docs.github.com/en/github/administering-a-repository/managing-repository-settings/managing-git-lfs-objects-in-archives-of-your-repository
However, including LFS files in repository archives makes the downloads of
archives count towards the repository organization's limit:

> If you choose to include Git LFS objects in archives of your repository, every download of those archives will count towards bandwidth usage for your account. Each account receives 1 GB per month of bandwidth for free, and you can pay for additional usage.

(I do not know what, if any, limit on download bandwidth GitHub has on repository downloads, e.g., release assets or the repository as a ZIP file or tarball. I have never encountered such a limit, so it does not seem to be of concern for testing on GitHub Actions (given that we download tarballs during the CI build from GitHub repositories, e.g., from the repository of `slugs`).)

(Also, my understanding is that in GitHub's documentation, the word "bandwidth" refers to cummulative amount of information transferred, and not to the rate of information transfer.)

(Sidenote: Git LFS uploads to GitHub's LFS storage do not count towards the bandwidth limit:
https://docs.github.com/en/github/managing-large-files/versioning-large-files/about-storage-and-bandwidth-usage#tracking-storage-and-bandwidth-use
)

Also, it seems to me that GitHub LFS can be affected by others, without write-access to the organization `tulip-control`, based on this:
https://docs.github.com/en/github/managing-large-files/versioning-large-files/about-storage-and-bandwidth-usage#tracking-storage-and-bandwidth-use

> Bandwidth and storage usage only count against the repository owner's quotas. In forks, bandwidth and storage usage count against the root of the repository network. Anyone with write access to a repository can push files to Git LFS without affecting their personal bandwidth and storage quotas or purchasing data packs. Forking and pulling a repository counts against the parent repository's bandwidth limit.

This part of the documentation seems to imply that whenever anyone clones the `data` repository, the Git LFS files downloaded from GitHub's LFS storage will count towards the (1 GB) limit of bandwidth per month for the organization `tulip-control`. Also, all Git LFS-related activity in forks of the repository `data` will count towards the monthly bandwidth quota for the organization `tulip-control`.


## Deleting old binary files from GitHub's LFS storage

Deleting old files stored in GitHub's LFS storage (i.e., old binaries) requires deleting and recreating the entire repository (in this case it would be the repository `data`), as described in GitHub's documentation:
https://docs.github.com/en/github/managing-large-files/versioning-large-files/removing-files-from-git-large-file-storage#git-lfs-objects-in-your-repository

Motivation for deleting outdated binary files:

- these files are unused, will not be used in the future (e.g., they are for old versions of an operating system), and keep storage space that could be used for other purposes

- GitHub has a limit on (free) bandwidth and storage of LFS files, as discussed above.


## Bitbucket support for deleting Git LFS files from Bitbucket storage

Bitbucket appears to support the deletion of files from their storage, without the need to delete (and recreate) the entire repository:
https://www.atlassian.com/git/tutorials/git-lfs#deleting-remote-files
I do not know about the bandwidth and file storage limitations of bitbucket regarding Git LFS files.


## `git` LFS supports only HTTPS

Recently, GitHub [discontinued password-based authentication](https://github.blog/2021-08-16-securing-your-github-account-two-factor-authentication/#no-more-password-based-authentication-for-git-operations). [`git-lfs` supports HTTPS/HTTP](https://github.com/git-lfs/git-lfs/wiki/Tutorial#credential-cachingstorage), so I think that using an SSH key is not possible. Some configuration that does not use a user password for HTTPS would be needed. I mention this detail for completeness.
